backend/models/comment.js

```diff
@@ -0,0 +1,121 @@
+/**
+ *	Copyright (C) 2019 3D Repo Ltd
+ *
+ *	This program is free software: you can redistribute it and/or modify
+ *	it under the terms of the GNU Affero General Public License as
+ *	published by the Free Software Foundation, either version 3 of the
+ *	License, or (at your option) any later version.
+ *
+ *	This program is distributed in the hope that it will be useful,
+ *	but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *	GNU Affero General Public License for more details.
+ *
+ *	You should have received a copy of the GNU Affero General Public License
+ *	along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+"use strict";
+
+const responseCodes = require("../response_codes.js");
+const utils = require("../utils");
+
+const fieldTypes = {
+	"action": "[object Object]",
+	"comment": "[object String]",
+	"consequence": "[object Number]",
+	"created": "[object Number]",
+	"guid": "[object Object]",
+	"from": "[object String]",
+	"likelihood": "[object Number]",
+	"mitigation": "[object String]",
+	"owner": "[object String]",
+	"pinPosition": "[object Array]",
+	"rev_id": "[object Object]",
+	"sealed": "[object Boolean]",
+	"to": "[object String]",
+	"viewpointGuid": "[object Object]"
+};
+
+class CommentGenerator {
+	constructor(owner, revId = undefined) {
+		this.guid = utils.generateUUID();
+		this.created = (new Date()).getTime();
+		this.owner = owner;
+
+		if (revId) {
+			if ("[object String]" === Object.prototype.toString.call(revId)) {
+				revId = utils.stringToUUID(revId);
+			}
+
+			this.rev_id = revId;
+		}
+	}
+}
+
+class TextCommentGenerator extends CommentGenerator {
+	constructor(owner, revId, commentText, viewpointGuid, pinPosition) {
+		super(owner, revId);
+		if (fieldTypes.comment === Object.prototype.toString.call(commentText)) {
+			if (commentText.length > 0 || fieldTypes.viewpointGuid === Object.prototype.toString.call(viewpointGuid)) {
```

Comment:
@Charence  
Does this mean I can still generate a comment with no comment and no screenshot if I provide you a viewpoint?

I guess it's up to the caller to check since you have no reference of the viewpoint data. But we probably need to check somewhere

Reply:
@carmenfan Yes >_<